compiler-k2/src/main/kotlin/com/squareup/anvil/compiler/k2/fir/merging/AnvilFirAnnotationMergingExtension.kt

```diff
@@ -0,0 +1,229 @@
+          newAnnotationCallPsi?.toKtPsiSourceElement(KtFakeSourceElementKind.PluginGenerated)
+            ?: componentAnnotation.source?.fakeElement(KtFakeSourceElementKind.PluginGenerated)
+        source = newSource
+
+        argumentList = buildArgumentList {
+          source = componentAnnotation.source?.fakeElement(KtFakeSourceElementKind.PluginGenerated)
+
+          arguments += buildNamedArgumentExpression {
+            name = Names.modules
+            isSpread = false
+            expression = buildArrayLiteral {
+              argumentList = buildArgumentList {
+                arguments += oldModules
+                arguments += contributedModules.map { it.toGetClassCall() }
+              }
+            }
+          }
+
+          componentAnnotation.argumentAt(
+            Names.dependencies,
+            index = 2,
+            unwrapNamedArguments = false,
+          )?.let {
+            arguments += it
+          }
+        }
+      },
+    )
+
+    return listOf()
+  }
+
+  private fun buildNewAnnotationPsi(
+    oldAnnotationCall: KtAnnotationEntry,
+    mergedModules: List<FqName>,
+  ): KtAnnotationEntry {
+
+    val oldAnnotationArguments = oldAnnotationCall.valueArgumentList
+      ?.arguments
+      .orEmpty()
+
+    // `modules = [SomeModule::class]`
+    val oldModulesArg = oldAnnotationArguments
+      // `Component` is a Java annotation with default argument values,
+      // so its arguments can be missing or in any order, but they must be named if they're present.
+      .firstOrNull { arg ->
+        val name = arg.getArgumentName()
+        name == null || name.text == "modules"
+      }
+
+    // `SomeModule::class, SomeOtherModule::class`
+    val existingModuleArgExpressions =
+      (oldModulesArg?.getArgumentExpression() as? KtCollectionLiteralExpression)
+        ?.innerExpressions
+        ?.map { it.text }
+        .orEmpty()
+
+    val imports = oldAnnotationCall.containingKtFile.importDirectives
+      .associate { imp ->
+        val fqName = imp.importedReference?.text
+          ?: error("import directive doesn't have a reference? $imp")
+
+        val name = imp.aliasName ?: imp.importedReference?.text?.substringAfterLast('.')
+          ?: error("import directive doesn't have a reference or alias? ${imp.text}")
+
+        fqName to name
+      }
+
+    val newModulesMaybeImported = mergedModules.map { it.asString() }
+      .map { moduleFqName ->
+        imports[moduleFqName] ?: moduleFqName
+      }
+
+    val allClassArgs = existingModuleArgExpressions
+      .plus(newModulesMaybeImported.map { "$it::class" })
+      .distinct()
+
+    val factory = oldAnnotationCall.ktPsiFactory()
+
+    val classArgList = allClassArgs.joinToString(separator = ", ")
+
+    val newModulesText = when {
+      oldModulesArg == null -> "modules = [$classArgList]"
+      existingModuleArgExpressions.isEmpty() -> "modules = [$classArgList]"
+      else -> "modules = [$classArgList]"
+    }
```

Comment:
When I first wrote this, it actually did different things for different conditions.  Now, it seems decidedly less useful.

Reply:
Good catch! Updated to be a direct assignment